Elemwise opt

aalmah opened the pull request:
- fixes #2160 
- made corrections and added some tests

nouiz commented:
travis is failing with this error (flake8, pep8):

```
======================================================================

FAIL: Test if flake8 is respected.

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest

    self.test(*self.arg)

  File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tests/test_flake8.py", line 274, in test_format_flake8

    raise AssertionError("FLAKE8 Format not respected")

AssertionError: FLAKE8 Format not respected

-------------------- >> begin captured stdout << ---------------------

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:1491:14: E127 continuation line over-indented for visual indent

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:1495:14: E127 continuation line over-indented for visual indent

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:1512:14: E127 continuation line over-indented for visual indent

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:1529:14: E127 continuation line over-indented for visual indent

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:3142:13: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:3149:1: E265 block comment should start with '# '

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:3150:1: E265 block comment should start with '# '

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:3162:19: E231 missing whitespace after ';'

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:3162:19: E702 multiple statements on one line (semicolon)

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4066:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4070:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4074:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4081:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4087:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4093:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4099:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4105:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4111:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4121:9: E129 visually indented line with same indent as next logical line

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:4130:9: E129 visually indented line with same indent as next logical line

--------------------- >> end captured stdout << ----------------------
```

memimo commented on theano/tensor/opt.py:
```diff
@@ -3067,11 +3123,48 @@ def local_remove_switch_const_cond(node):
+    try:
+        shp = node.fgraph.shape_feature.shape_of[node.inputs[0]][node.op.i]
+        c = get_scalar_constant_value(shp)
+        import pdb;pdb.set_trace()
```
This should be removed.

memimo commented on theano/tensor/opt.py:
```diff
@@ -3936,6 +4029,110 @@ def local_elemwise_sub_zeros(node):
+
+    :note: Those case appear in the graph generated around scan.  This
+           don't remove much computation, but make the graph easier to
+           read.
```
Maybe this could be written more clearly. Also there are some grammar mistakes. Something like:

These cases appear in the graph generated by scan. This will not reduce computation but will make the graph easier to read.

memimo commented:
I made an initial review. I'm not very familiar with guts of theano, but just skimming through I couldn't find any major issue. 
For travis error about formatting issue you might want to use https://github.com/google/yapf which makes your life super easy autmaticly making your script PEP8 compatible.

nouiz commented:
Don't use that script in this PR, let Iban handle the conversion of each
file. thanks. I'll review more later.

On Tue, Jul 21, 2015 at 10:56 AM, Mehdi Mirza <notifications@github.com>
wrote:

> I made an initial review. I'm not very familiar with guts of theano, but
> just skimming through I couldn't find any major issue.
> For travis error about formatting issue you might want to use
> https://github.com/google/yapf which makes your life super easy
> autmaticly making your script PEP8 compatible.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3173#issuecomment-123358835>.
>

aalmah commented:
I don't understand what I need to do now for fixing travis errors. I was planning to fix problems in the opt.py file only, not sure if that will conflict with Iban's work.

nouiz commented:
That is the good thing to do. It should not conflict with anything.

You have other tests failing then stuff related to flake8 in travis.

On Tue, Jul 21, 2015 at 11:28 AM, Amjad Almahairi <notifications@github.com>
wrote:

> I don't understand what I need to do now for fixing travis errors. I was
> planning to fix problems in the opt.py file only, not sure if that will
> conflict with Iban's work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3173#issuecomment-123370409>.
>

nouiz commented:
Don't forget to make a comment when you update a PR, otherwise, we don't get a notification about this, so this delay our review of the update.

travis is failing with those errors;

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:2696:24: E126 continuation line over-indented for hanging indent

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:2790:17: E126 continuation line over-indented for hanging indent

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:2847:24: E126 continuation line over-indented for hanging indent

/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/tensor/opt.py:2863:24: E126 continuation line over-indented for hanging indent

aalmah commented:
- I have reverted the unnecessary flake8 changes and added just the ones that make test_flake8 test work. There is this file that still has flake8 failure, but this pull request has nothing to do with it:
Theano/theano/generated_version.py:7:9: E126 continuation line over-indented for hanging indent

- I also added a fix for FAST_COMPILE mode that was failing, in addition to some comments from Fred

nouiz commented:
There is one failure on travis:

```
======================================================================

FAIL: test_isclose (theano.tensor.tests.test_basic.test_comparison)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tensor/tests/test_basic.py", line 4194, in test_isclose

    dtype="bool"

AssertionError: False is not true : False

----------------------------------------------------------------------
```

can you check?

nouiz commented:
this PR need a rebase.

aalmah commented:
Rebase is done

dwf commented:
@aalmah @nouiz This PR looks good to me, I think it can be merged safely.

nouiz commented on theano/tensor/opt.py:
```diff
@@ -4128,6 +4208,109 @@ def local_elemwise_sub_zeros(node):
+       node.inputs[0].owner and \
+       isinstance(node.inputs[0].owner.op, Shape_i) and \
+       T.extract_constant(node.inputs[1]) == 0:
+        return [T.zeros_like(node.outputs[0])]
```
This is bad practice as it could create an infinit loop depending of graph optimization order.

use return zeros(like(node.inputs[0])

nouiz commented on theano/tensor/opt.py:
```diff
@@ -4128,6 +4208,109 @@ def local_elemwise_sub_zeros(node):
+       T.extract_constant(node.inputs[0]) == 0 and \
+       node.inputs[1].owner and \
+       isinstance(node.inputs[1].owner.op, Shape_i):
+        return [T.zeros_like(node.outputs[0])]
```
similar, but use return zeros(like(node.inputs[1])

nouiz commented on theano/tensor/tests/test_opt.py:
```diff
@@ -3135,6 +3135,156 @@ def test_local_fill_useless():
+        assert_eqs_const(f.maker.fgraph.toposort(), 0)
+
+        f = theano.function([x], T.le(x, x), mode=mode)
+        assert_eqs_const(f.maker.fgraph.toposort(), 1)
```
I'll make a new issue to optimize that case and GE case.

nouiz replied:
https://github.com/Theano/Theano/issues/3370

nouiz commented on theano/tensor/tests/test_opt.py:
```diff
@@ -3135,6 +3135,156 @@ def test_local_fill_useless():
+
+    def assert_identity(self, f):
+        topo = f.maker.fgraph.toposort()
+        assert topo[0].op == deep_copy_op
```
Assert that the len of the toposort is 1

nouiz commented on theano/tensor/tests/test_opt.py:
```diff
@@ -3135,6 +3135,156 @@ def test_local_fill_useless():
+        x_val = 127
+        assert f(x_val) == x_val
+        x_val = numpy.random.randint(255)-128
+        assert f(x_val) == x_val
```
with that extra assert, just keep one assert with value. All the rest would be duplicate.

nouiz replied:
and remove the last case.

nouiz commented on theano/tensor/opt.py:
```diff
@@ -4128,6 +4208,109 @@ def local_elemwise_sub_zeros(node):
+    # Elemwise[maximum](X.shape[i], 0) -> X.shape[i]
+    # Elemwise[maximum](0, X.shape[i]) -> X.shape[i]
+    # Elemwise[minimum](X.shape[i], 0) -> 0
+    # Elemwise[minimum](0, X.shape[i]) -> 0
```
uncomment those 4 cases. They are optimized. They need tests

nouiz commented on theano/tensor/opt.py:
```diff
@@ -4128,6 +4208,109 @@ def local_elemwise_sub_zeros(node):
+    # Comparing to itself is constant
+    Elemwise[{LT,GT}](X, X) -> Elemwise[zeros](X)
+    Elemwise[{LE,GE}](X, X) -> Elemwise[ones](X)
+    Elemwise[{minimum,maximum}](X, X) -> X
```
those aren't tested.

nouiz commented on theano/tensor/opt.py:
```diff
@@ -4128,6 +4208,109 @@ def local_elemwise_sub_zeros(node):
+
+    # The shape can be replaced with sum of shapes
+    Elemwise[LT](sum([anything that is shapes]), 0) -> Elemwise[zeros](X)
+    Elemwise[GE](sum([anything that is shapes]), 0) -> Elemwise[ones](X)
```
idem, no tests.

nouiz replied:
rename sum to add.

nouiz commented on theano/tensor/opt.py:
```diff
@@ -4128,6 +4208,109 @@ def local_elemwise_sub_zeros(node):
+    if isinstance(node.op.scalar_op, scalar.LT) and \
+       node.inputs[0].owner and \
+       isinstance(node.inputs[0].owner.op, Elemwise) and \
+       isinstance(node.inputs[0].owner.op, scalar.Add) and \
```
should be ...owner.op.scalar_op, idem other opt bellow

aalmah commented:
I have addressed all of the issues you mentioned. Please take a look and let me know if there's anything more to add.

nouiz commented:
I think my last comment is the only fix left.

nouiz commented:
Also, 2 tests are failing in travis:

```
======================================================================

FAIL: test_shape_add_inequality (theano.tensor.tests.test_opt.Test_local_useless_elemwise_comparison)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tensor/tests/test_opt.py", line 3283, in test_shape_add_inequality

    self.assert_eqs_const(f, 0)

  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tensor/tests/test_opt.py", line 3209, in assert_eqs_const

    assert len(topo) == 1, topo

AssertionError: [Shape(y), Shape(x), Subtensor{int64}(Shape.0, Constant{0}), Subtensor{int64}(Shape.0, Constant{0}), Elemwise{add,no_inplace}(Subtensor{int64}.0, Subtensor{int64}.0), Elemwise{lt,no_inplace}(Elemwise{add,no_inplace}.0, TensorConstant{0})]

======================================================================

FAIL: test_shape_inequality_with_self (theano.tensor.tests.test_opt.Test_local_useless_elemwise_comparison)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tensor/tests/test_opt.py", line 3251, in test_shape_inequality_with_self

    self.assert_eqs_const(f, 0)

  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tensor/tests/test_opt.py", line 3209, in assert_eqs_const

    assert len(topo) == 1, topo

AssertionError: [Shape(x), Subtensor{int64}(Shape.0, Constant{0}), Elemwise{lt,no_inplace}(Subtensor{int64}.0, TensorConstant{0})]

----------------------------------------------------------------------

```

aalmah commented:
@nouiz I have changed the dtype and fixed the tests
